### PR TITLE
Use ClusterState as Consistency Source for Snapshot Repositories

### PR DESCRIPTION
**Pull request by original-brownbear**
Follow up to #49729 

This change removes falling back to listing out the repository contents to find the latest `index-N` in write-mounted blob store repositories.
This saves 2-3 list operations on each snapshot create and delete operation. Also it makes all the snapshot status APIs cheaper (and faster) by saving one list operation there as well in many cases.
This removes the resiliency to concurrent modifications of the repository as a result and puts a repository in a `corrupted` state in case loading `RepositoryData` failed from the assumed generation.

Note to reviewers:

I suggest starting from the new test ``CorruptedBlobStoreRepositoryIT` illustrating the mechanics of a repository being marked as "corrupted" to make sure we're in agreement on the behavior change here. I know we discussed in the snapshot meeting a while back, but let's make sure we're all in agreement on the tradeoffs made here:

* (good) Slightly faster and cheaper (cloud repos) repository operations (both dimensions are somewhat irrelevant I'd say for most users)

* (good-neutral) Concurrent modifications to the repository by multiple clusters are detected much more reliably and force the user to re-add the repo by hand, making it so that the dangerous situation of having two clusters write to the repo at the same time is impossible. This seems like mostly a win, but he off situation where a user may use the repo for writes across two clusters and manually ensures that operations don't run at the same time would be broken by this change

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1

**Comment by original-brownbear** [comment]
Need to get https://github.com/elastic/elasticsearch/pull/49514 in first to stabilize `SnapshotResiliencyTests` enough so that they don't become unstable from this change

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/2

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 1]
```diff
@@ -279,29 +294,44 @@ protected void doClose() {
...
-        }
-        final RepositoryCleanupInProgress cleanupInProgress = state.custom(RepositoryCleanupInProgress.TYPE);
-        if (bestGenerationFromCS == RepositoryData.EMPTY_REPO_GEN && cleanupInProgress != null) {
-            bestGenerationFromCS = bestGeneration(cleanupInProgress.entries());
-        }
-
         metadata = getRepoMetaData(state);
-        final long finalBestGen = Math.max(bestGenerationFromCS, metadata.generation());
-        latestKnownRepoGen.updateAndGet(known -> Math.max(known, finalBestGen));
+        if (bestEffortConsistency) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: This block can go away in `master` after backport to `7.x`, I'd keep the "old"-way of guessing from the in-progress actions for `7.x` though.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 2]
```diff
@@ -985,36 +1020,116 @@ public void endVerification(String seed) {
 
     // Tracks the latest known repository generation in a best-effort way to detect inconsistent listing of root level index-N blobs
     // and concurrent modifications.
-    private final AtomicLong latestKnownRepoGen = new AtomicLong(RepositoryData.EMPTY_REPO_GEN);
+    private final AtomicLong latestKnownRepoGen = new AtomicLong(RepositoryData.UNKNOWN_REPO_GEN);
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Can't just start from the empty generation any more because we're not per-default falling back to listing.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 3]
```diff
@@ -985,36 +1020,116 @@ public void endVerification(String seed) {
...
-        ActionListener.completeWith(listener, () -> {
+        if (latestKnownRepoGen.get() == RepositoryData.CORRUPTED_REPO_GEN) {
+            listener.onFailure(corruptedStateException(null));
+            return;
+        }
+        if (bestEffortConsistency == false && latestKnownRepoGen.get() == RepositoryData.UNKNOWN_REPO_GEN) {
+            if (clusterService.state().nodes().isLocalNodeElectedMaster() == false) {
+                // Non-master node does not load RepositoryData before master has at least loaded it once and thus initialized in the
+                // cluster state
+                listener.onFailure(new RepositoryException(metadata.name(), "Repository not initialized yet"));
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Non-master nodes only ever load `RepositoryData`  in `org.elasticsearch.index.shard.StoreRecovery#restore` (in which case master will have loaded `RepositoryData` before to verify that the snapshot to be restored actually exists in the repo) as far as I can see so in the real world we should never get here (except for some freak concurrency cases where master fails over concurrently).

File: test/framework/src/main/java/org/elasticsearch/repositories/AbstractThirdPartyRepositoryTestCase.java [hunk 4]
```diff
@@ -76,6 +76,7 @@ public void setUp() throws Exception {
     @Override
     public void tearDown() throws Exception {
         deleteAndAssertEmpty(getRepository().basePath());
+        client().admin().cluster().prepareDeleteRepository("test-repo").get();
```
Comment (original-brownbear) [inline test/framework/src/main/java/org/elasticsearch/repositories/AbstractThirdPartyRepositoryTestCase.java]: Can't reuse the repo across tests that clear out its content any longer

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 5]
```diff
@@ -985,36 +1020,116 @@ public void endVerification(String seed) {
...
+            } else {
+                // After initializing the repo in the cluster state we retry loading the latest RepositoryData on the generic pool
+                initializeRepoInClusterState(ActionListener.wrap(v ->
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: I don't think we should initialize this here (as it can be based on an inconsistent read). Instead I prefer we add this after a successful write repo operation and load leniently until that has happened.

**Review by ywelsch — CHANGES_REQUESTED** [review]
Thanks for tackling this, Armin. I've left some comments.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 5 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Thanks, that's a sweet idea I think. That way we force at least one more listing before writing the new repo generation and are even safer in a freaky scenario where the repo was just modified and then mounted for writing :) Will adjust accordingly.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 3 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: This actually becomes a non-issue thanks to your suggestion in https://github.com/elastic/elasticsearch/pull/49060#discussion_r357232276 since we can simply fall back to listing on the non-master nodes so long as the master hasn't initialized the latest generation in the CS

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 6]
```diff
@@ -279,29 +294,44 @@ protected void doClose() {
     // #latestKnownRepoGen if a newer than currently known generation is found
     @Override
     public void updateState(ClusterState state) {
-        if (readOnly) {
+        bestEffortConsistency = isReadOnly() || state.nodes().getMinNodeVersion().before(RepositoryMetaData.REPO_GEN_IN_CS_VERSION);
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: You're right the upgrade experience is potentially broken here. I'll fix this together with the adjustments to the initialization.

> Also, I wonder if we should (at least in 7.x) allow a feature flag to disable enforcing consistency. At least to ensure we are not breaking existing use cases (e.g. something fishy on Cloud)

That sounds like a good idea. I wonder how to implement this though: should we do a system property? (then things could get a little weird if it's not set on all nodes but it seems like the easiest approach)

[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: I would prefer a deprecated cluster setting

[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Cluster setting it is -> will do :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 7]
```diff
@@ -279,29 +294,44 @@ protected void doClose() {
...
+            }
+            final long finalBestGen = Math.max(bestGenerationFromCS, metadata.generation());
+            latestKnownRepoGen.updateAndGet(known -> Math.max(known, finalBestGen));
+        } else {
+            if (metadata.generation() == RepositoryData.UNKNOWN_REPO_GEN) {
+                // We break out here, assuming this was the result of moving from a mixed cluster that didn't track the generation in the
+                // cluster state to a cluster that does track the generation in the cluster state
+                return;
+            }
+            final long previousBest = latestKnownRepoGen.getAndSet(metadata.generation());
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: My thinking was, that if we get here the cluster state is messed up in some form and we can't really change that anyway (it would be a bug in some cluster state update before) and I simply added this to have the debug logging + assertion. If we don't update the generation pointer on a node while the CS was incorrectly updated in some way we don't get anything out of that do we?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 5 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Removed the initialization logic in the read path completely now and simply made `BlobStoreRepository` use the old style listing fallback as long as the cluster state contains `RepositoryData#UNKNOWN_REPO_GEN` in the `RepositoryMetaData`. That also solves all the BwC issues IMO, since if we fall over from an older master, the cluster state will container `RepositoryData#UNKNOWN_REPO_GEN` and if we're already on a new version master then the contents of the cluster state will be correct if that master has ever written a new `index-N` or will be `RepositoryData#UNKNOWN_REPO_GEN` otherwise.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: See my comment here https://github.com/elastic/elasticsearch/pull/49060#discussion_r357605538  on the BwC situation, that should be resolved now.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 8]
```diff
@@ -985,36 +1020,116 @@ public void endVerification(String seed) {
...
     // and concurrent modifications.
-    private final AtomicLong latestKnownRepoGen = new AtomicLong(RepositoryData.EMPTY_REPO_GEN);
+    private final AtomicLong latestKnownRepoGen = new AtomicLong(RepositoryData.UNKNOWN_REPO_GEN);
 
     @Override
     public void getRepositoryData(ActionListener<RepositoryData> listener) {
-        ActionListener.completeWith(listener, () -> {
-            // Retry loading RepositoryData in a loop in case we run into concurrent modifications of the repository.
-            while (true) {
+        if (latestKnownRepoGen.get() == RepositoryData.CORRUPTED_REPO_GEN) {
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: This is a non-issue now I think since with the removal of the initialization logic we're only really using the value once before we enter the read loop that spins on this value and actively expects and deals with it changing.

**Comment by original-brownbear** [comment]
@ywelsch thanks for taking a look. I addressed all your comments now I hope. I removed the repo initialization from the read path and added a deprecated setting to allow concurrent repo modifications like we currently have (see the new test added to `CorruptedBlobStoreRepositoryIT`).
I think this is good for another review :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 9]
```diff
@@ -184,6 +185,14 @@
      */
     public static final Setting<Boolean> COMPRESS_SETTING = Setting.boolSetting("compress", true, Setting.Property.NodeScope);
 
+    /**
+     * When set to {@code true}, {@link #bestEffortConsistency} will be set to {@code true} and concurrent modifications of the repository
+     * contents will not result in the repository being marked as corrupted.
+     * Note: This setting is intended as a backwards compatibility solution for 7.x and will go away in 8.
+     */
+    public static final Setting<Boolean> ALLOW_CONCURRENT_MODIFICATION =
+        Setting.boolSetting("allow_concurrent_mod", false, Setting.Property.Deprecated, Setting.Property.Dynamic);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Let's fully spell out the name, i.e. `allow_concurrent_modifications`

Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: Also, do we need to define NodeScope but not dynamic, like we have for COMPRESS_SETTING?

**Review by ywelsch — COMMENTED** [review]
I wonder if we should always be lenient w.r.t. to loading repository data when `pendingGeneration != safeGeneration`, in case where the pendingGeneration might have actually been successfully written to the repo (and possibly even index.latest been updated). It's possible that other read-only repos will observe that new state of the repo, and the writer to the repo will then bring it backwards (in particular risky if cluster state updates are slow at that particular point in time, and there is a longer lag between writing to the repo and updating safe repository).

**Comment by original-brownbear** [comment]
>  It's possible that other read-only repos will observe that new state of the repo, and the writer to the repo will then bring it backwards (in particular risky if cluster state updates are slow at that particular point in time, and there is a longer lag between writing to the repo and updating safe repository).

Bringing the repo backwards is not really a possibiilty I think because if the cluster state update after the write failed it will be retried by the next master that will just finalize the still ongoing operation. That's exactly what was fixed in https://github.com/elastic/elasticsearch/pull/49995 I'd say. There we had tests failing because the repo was "taken back" because the snapshot job was removed from the CS without retrying the write.

If we add leniency here, it complicates things quite a bit IMO because we get situations like:

0. We have some hanging master that is stuck before writing index-N and just failed
1. Master A now tries to read non-existent index-N and destroys read-after-write consistency for that key then fails over for some reason
2. The hanging master comples the write to index-N
3. Whatever new master is elected may or may not see the just written `index-N` at the pending generation

(this would also fail existing tests for eventually consistent blob stores that explicitly make sure we never read from an inconsistent key)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 9 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: > Also, do we need to define NodeScope but not dynamic, like we have for COMPRESS_SETTING?

Not really, all these attributes are just descriptive only as far as I can tell since these are not registered settings but and simply applied to the repository metadata. I'll remove the `Dynamic` part as it's pointless :)

**Comment by ywelsch** [comment]
> it will be retried by the next master that will just finalize the still ongoing operation

probably only unless you have a full cluster restart. In that case, SnapshotInProgress and friends are lost whereas RepositoriesMetaData is not.

I want to mainly understand the worst case here, and how that could possibly affect a user. In theory we could fix the concurrent master issue by appending the master term as an additional tie breaker to the file name.

**Comment by original-brownbear** [comment]
> I want to mainly understand the worst case here, and how that could possibly affect a user. In theory we could fix the concurrent master issue by appending the master term as an additional tie breaker to the file name.

I think the worst case is as you mention the full cluster restart right before a snapshot finishes (i.e. before `RepositoryMetadata` generations get aligned). In that case, if you were to restore last snapshot taken and concurrently ran either a repository cleanup or snapshot delete the restore may fail mid-way because the cleanup or delete will interpret the last snapshot's files as unreferenced and remove them. 
I feel like this is a really fringe case though. Right now you have a bunch of failure scenarios in various ES versions when restoring while deleting from a repository (the step of reading `RepositoryData` which happens on the data nodes restoring is error prone in any version if a new one is written concurrently ... ) so I'm not sure it's worth it worrying about this? It's super unlikely and even without this change deleting anything while restoring has the potential to kill the restore, we're just adding one new very unlikely scenario.

**Comment by original-brownbear** [comment]
> I want to mainly understand the worst case here, and how that could possibly affect a user. In theory we could fix the concurrent master issue by appending the master term as an additional tie breaker to the file name.

I think the worst case is as you mention the full cluster restart right before a snapshot finishes (i.e. before `RepositoryMetadata` generations get aligned). In that case, if you were to restore last snapshot taken and concurrently ran either a repository cleanup or snapshot delete the restore may fail mid-way because the cleanup or delete will interpret the last snapshot's files as unreferenced and remove them. 
I feel like this is a really fringe case though. Right now you have a bunch of failure scenarios in various ES versions when restoring while deleting from a repository (the step of reading `RepositoryData` which happens on the data nodes restoring is error prone in any version if a new one is written concurrently ... ) so I'm not sure it's worth it worrying about this? It's super unlikely and even without this change deleting anything while restoring has the potential to kill the restore, we're just adding one new very unlikely scenario.

> appending the master term as an additional tie breaker to the file name

So basically we'd have `index-{N}-{term}` and if we run into say pending generation `5`, we'd list `index-5-*` and move to that as the safe generation if we were to find a blob for `index-5-{someTerm}`? I think we could do that, but since it really only helps the concurrent repository access case I'm not sure it's worth the changes (we'd have to change what we store in `index.latest` and refactor the generation to be two `long` or a `String` wouldn't we?).

**Comment by original-brownbear** [comment]
@ywelsch as discussed earlier today, I added logic to identify a "dirty" startup of the repository now in e3f7ff2 (code and test) and 564202b (docs) that shoul make us never lose the latest state (assuming the list operation works well enough on full cluster restart :) but it should be fine and even if it isn't once in a blue moon, the repo can not become corrupted as a result still).

**Review by ywelsch — APPROVED** [review]
LGTM. Can you, in a follow-up, also add more tests for read-only repositories, given that they have a slightly different way to load stuff?

**Comment by original-brownbear** [comment]
Thanks Yannick!

> Can you, in a follow-up, also add more tests for read-only repositories, given that they have a slightly different way to load stuff?

Will do :)
